dbms/src/Dictionaries/HashedDictionary.cpp

```diff
@@ -3,6 +3,23 @@
 #include "DictionaryBlockInputStream.h"
 #include "DictionaryFactory.h"
 
+namespace
+{
+
+/// google::sparse_hash_map
+template <class T> auto first(const T &lhs) -> decltype(lhs.first)
```

Comment:
Wow, this is my first time seeing the trailing return type syntax. You're not converting the return value to a different type, so it doesn't add anything over just `auto`, and can be omitted.

Reply:
Nope, this is SFINAE, well not exactly, but the same purpose.

`HashMap` provides:
- `getFirst()`
- `getSecond()`

`google::sparse_hash_map` provides:
- `.first`
- `.second`

So we need to use  `getFirst()`/`getSecond()` only if we have it otherwise `.first`/`.second`

Reply:
Ok. BTW, we cannot tolerate such tricks without a clarifying comment in code (added).